services/cronback-project-srv/handler.rs

```diff
@@ -0,0 +1,176 @@
+        match status {
+            | Some(st) => {
+                Ok(Response::new(GetProjectStatusResponse {
+                    status: st.into(),
+                }))
+            }
+            | None => {
+                Err(ProjectStoreHandlerError::NotFound(format!(
+                    "{}",
+                    project_id
+                )))?
+            }
+        }
+    }
+
+    async fn set_project_status(
+        &self,
+        request: Request<SetProjectStatusRequest>,
+    ) -> Result<Response<SetProjectStatusResponse>, Status> {
+        let (_, _, req) = request.into_parts();
+        let project_id: ProjectId = req.id.unwrap().into();
+        let project_id = project_id
+            .validated()
+            .map_err(|e| Status::invalid_argument(e.to_string()))?;
+
+        let old_status = self
+            .project_store
+            .get_status(&project_id)
+            .await
+            .map_err(ProjectStoreHandlerError::Store)?;
+
+        let Some(old_status) = old_status else {
+            return Err(ProjectStoreHandlerError::NotFound(
+                project_id.to_string(),
+            )
+            .into());
+        };
+
+        self.project_store
+            .set_status(&project_id, req.status.into())
+            .await
+            .map_err(ProjectStoreHandlerError::Store)?;
+
+        // NOTE: The `old_status` here might not always be 100% accurate as it
+        // might race with another update.
+        e!(
+            project_id = project_id,
+            ProjectStatusUpdated {
+                old_status: old_status.into(),
+                new_status: req.status,
+            }
+        );
+
+        Ok(Response::new(SetProjectStatusResponse {
+            status: req.status,
+        }))
+    }
+
+    async fn is_project_exists(
+        &self,
+        request: Request<IsProjectExistsRequest>,
+    ) -> Result<Response<IsProjectExistsResponse>, Status> {
+        let (_, _, req) = request.into_parts();
+        let project_id: ProjectId = req.id.unwrap().into();
+        let project_id = project_id
+            .validated()
+            .map_err(|e| Status::invalid_argument(e.to_string()))?;
+        let exists = self
+            .project_store
+            .exists(&project_id)
+            .await
+            .map_err(ProjectStoreHandlerError::Store)?;
+        Ok(Response::new(IsProjectExistsResponse { exists }))
+    }
+}
+
+#[derive(Error, Debug)]
+pub(crate) enum ProjectStoreHandlerError {
+    #[error("Project '{0}' is unknown to the store!")]
+    NotFound(String),
+    #[error("Operation on underlying database failed: {0}")]
+    Store(#[from] DatabaseError),
+}
+
+impl From<ProjectStoreHandlerError> for Status {
+    fn from(e: ProjectStoreHandlerError) -> Self {
+        // match variants of TriggerError
+        match e {
+            | ProjectStoreHandlerError::NotFound(e) => Status::not_found(e),
+            | e => Status::invalid_argument(e.to_string()),
```

Comment:
So, a `DatabaseError` is a gRPC `invalid_argument`? Is that your intention here?

**Side Note:** I'd expect the default/blanket match on error to map to a more generic/general-purpose error like `Status::internal`. And to do that, we need to have an explicit map for the rest of  _logical_ errors to the right gPRC status code.

Reply:
Good catch!

Reply:
Hmmm, it seems like we're doing that in both the scheduler and dispatcher as well. Should I modify them to all return `Status::unavailable` or `Status::internal`?

Reply:
Scheduler has a correct but error-prone implementation. The rest of the errors are mapped correctly to `invalid_argument` because `TriggerError` doesn't have blanket/transparent error bypass—`#[from] ...`. It's probably better to have exhaustive match to avoid mislabeling new errors though.

Reply:
@MohamedBassem Also moving this to `error.rs` at service top-level package is a good practice to make it visible where do we handle the gRPC handler errors. See cronback-scheduler-srv/src/error.rs for inspiration.